DIPs/DIP27.md

```diff
@@ -0,0 +1,172 @@
+# First class functions
+
+| Section        | Value                                  |
+|----------------|----------------------------------------|
+| DIP:           | 27                                     |
+| Author:        | Amaury SÉCHET                          |
+| Status:        |                                        |
+
+## Abstract
+
+This DIP is the first of 3. It intends to define what is a function in D. The DIP propose a radically simple approach by deciding that all function are first class function. The DIP proposes ways to mitigate breakage involved and allow optional parenthesis when no confusion is possible.
+
+## Description
+
+Function are still defined with the same syntax :
+
+```d
+ReturnType functionName(Parameters) {
+  // Function body.
+}
+```
+
+However, this is now strictly equivalent to :
+
+```d
+enum functionName = ReturnType function(Parameters) {
+  // Function body.
+};
+```
+
+The function has a function type and can be used as this anywhere a function is expected :
+
+```d
+static assert(is(typeof(functionName) == ReturnType function(Parameters)); // OK
+
+ReturnType function(Parameters) foo = functionName; // OK
+
+void buzz(ReturnType function(Parameters) qux);
+buzz(foo); // OK
+buzz(bar); // OK
+
+auto a = functionName;
+static assert(is(typeof(a) == ReturnType function(Parameters)); // OK
+
+auto b = functionName();
+static assert(is(typeof(b) == ReturnType); // OK
+```
+
+Note that what is above is simply consequences of the simple rule expressed above.
+
+### Is Expression
+
+The function type specialization used in is expression has to be modified as well. It does match the first class function we defined above, and alias parameters as defined on https://dlang.org/spec/expression.html#IsExpression .
+
+```d
+void foo() {}
+auto bar = foo;
+
+static assert(is(typeof(foo) == function)); // OK
+static assert(is(typeof(bar) == function)); // OK
```

Comment:
Indeed, I need to add a blurb about this. IMO, this is a feature, not a bug, considering there were bugs related to this even in phobos (getopt for instance, would not accept firstc lass functions at the time the DIP was written, no idea if it was fixed). When nobody gets it right, including phobos's contrbutors, I think it is time to admit this was a bug, not a feature.